presto-main/src/main/java/com/facebook/presto/sql/analyzer/TypeSignatureProvider.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.analyzer;
+
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeManager;
+import com.facebook.presto.spi.type.TypeSignature;
+import com.facebook.presto.util.ImmutableCollectors;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class TypeSignatureProvider
+{
+    private final boolean hasDependency;
+    private final Function<List<Type>, TypeSignature> typeSignatureSupplier;
+
+    public TypeSignatureProvider(TypeSignature typeSignature)
+    {
+        this.hasDependency = false;
+        this.typeSignatureSupplier = ignored -> typeSignature;
+    }
+
+    public TypeSignatureProvider(Function<List<Type>, TypeSignature> typeSignatureSupplier)
+    {
+        this.hasDependency = true;
+        this.typeSignatureSupplier = requireNonNull(typeSignatureSupplier, "typeSignatureSupplier is null");
+    }
+
+    public boolean hasDependency()
+    {
+        return hasDependency;
+    }
+
+    public TypeSignature getTypeSignature()
+    {
+        checkState(!hasDependency);
```

Comment:
Doing the validation in the other function as you suggested would require having an integer in the type signature provider. I don't have such an integer. As a result, I don't do the check.

Given the check can be done in the `resolver`, I don't find it particularly important to have the number available so that the check can be done.